kernel/kernel-api/src/main/java/io/delta/kernel/internal/replay/CRCInfo.java

```diff
@@ -0,0 +1,65 @@
+/*
+ * Copyright (2024) The Delta Lake Project Authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.delta.kernel.internal.replay;
+
+import io.delta.kernel.data.ColumnarBatch;
+import io.delta.kernel.engine.Engine;
+import io.delta.kernel.internal.actions.Metadata;
+import io.delta.kernel.internal.actions.Protocol;
+import io.delta.kernel.types.StructType;
+
+public class CRCInfo {
+
+  public static CRCInfo fromColumnarBatch(
+      Engine engine, long version, ColumnarBatch batch, int rowId) {
+    // fromColumnVector already takes care of nulls
```

Comment:
metadata and protocol are required in the CRC. https://github.com/delta-io/delta/blob/master/PROTOCOL.md#version-checksum-file-schema

if they aren't there, then we should not yield a CRC info. futher, let's not just propagate nulls around our code base everwhere. that's what Optional is for in java.

why don't we have this `CRCInfo::fromColumnarBatch` return an Optional then, since it needs to do some complex poking around in the underlying vectors.

`fromColumnVector already takes care of nulls` and this is tech debt that I don't like at all :(

Reply:
> why don't we have this CRCInfo::fromColumnarBatch return an Optional then, since it needs to do some complex poking around in the underlying vectors.

+1 this would be an improvement (and we don't need to do the check for their presence elsewhere in the code)

Reply:
But let's still log if we find a crc file that's missing one of the two!

Reply:
Done